teuthology/repo_utils.py

```diff
@@ -114,25 +114,100 @@ def enforce_repo_state(repo_url, dest_path, branch, commit=None, remove_on_error
     # sentinel to track whether the repo has checked out the intended
     # version, in addition to being cloned
     repo_reset = os.path.join(dest_path, '.fetched_and_reset')
+    log.info("enforce_repo_state %s %s %s %s %s", dest_clone, dest_path, repo_url, branch, commit)
     try:
-        if not os.path.isdir(dest_path):
-            clone_repo(repo_url, dest_path, branch, shallow=commit is None)
+        if not os.path.isdir(dest_clone):
+            bare_repo(dest_clone)
         elif not commit and not is_fresh(sentinel):
-            set_remote(dest_path, repo_url)
-            fetch_branch(dest_path, branch)
+            #set_remote(dest_path, repo_url)
+            #fetch_branch(dest_path, branch)
             touch_file(sentinel)
 
-        if commit and os.path.exists(repo_reset):
-            return
+        #if commit and os.path.exists(repo_reset):
+            #return
 
-        reset_repo(repo_url, dest_path, branch, commit)
-        touch_file(repo_reset)
+        myfetch(dest_clone, repo_url, branch, commit)
+        myworkspace(dest_clone, dest_path)
+        #reset_repo(repo_url, dest_path, branch, commit)
+        #touch_file(repo_reset)
         # remove_pyc_files(dest_path)
     except (BranchNotFoundError, CommitNotFoundError):
         if remove_on_error:
             shutil.rmtree(dest_path, ignore_errors=True)
         raise
 
+def bare_repo(git_dir):
+    log.info("bare_repo %s", git_dir)
+    args = ['git', 'init', '--bare', git_dir]
+    proc = subprocess.Popen(args)
```

Comment:
If you don't need a lot of the complexity that Popen provides. I heartily recommend using the simpler `subprocess.run` function (avialable since 3.5).  You can use the `check=True` parameter to have it raise exceptions on non-zero exit automatically too.